capstone/src/main/java/com/google/sps/gmail/GmailAPILoader.java

```diff
@@ -0,0 +1,98 @@
+package com.google.sps.gmail;
+
+import com.google.api.client.auth.oauth2.Credential;
+import com.google.api.client.googleapis.auth.oauth2.GoogleCredential;
+import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
+import com.google.api.client.http.javanet.NetHttpTransport;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.api.services.gmail.Gmail;
+import com.google.common.collect.Streams;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.net.URLEncoder;
+import java.security.GeneralSecurityException;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.json.JSONObject;
+
+public class GmailAPILoader {
+  private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+  private static final String APPLICATION_NAME = "clubhub-step-2020";
+
+  public static Gmail getGmailService() throws IOException, GeneralSecurityException {
+    // Add credentials for using API
+    Credential authorize =
+        new GoogleCredential.Builder()
+            .setTransport(GoogleNetHttpTransport.newTrustedTransport())
+            .setJsonFactory(JSON_FACTORY)
+            .setClientSecrets(Secrets.CLIENT_ID, Secrets.CLIENT_SECRET)
+            .build()
+            .setRefreshToken(Secrets.REFRESH_TOKEN)
+            .setAccessToken(getAccessToken());
+
+    // Create Gmail API Service
+    final NetHttpTransport HTTP_TRANSPORT = GoogleNetHttpTransport.newTrustedTransport();
+    Gmail service =
+        new Gmail.Builder(HTTP_TRANSPORT, JSON_FACTORY, authorize)
+            .setApplicationName(APPLICATION_NAME)
+            .build();
+    return service;
+  }
+
+  private static String getAccessToken() throws IOException {
+    try {
+      // Add secret values to Map
+      Map<String, Object> params = new LinkedHashMap<>();
+      params.put("grant_type", "refresh_token");
+      params.put("client_id", Secrets.CLIENT_ID);
```

Comment:
Since this file is being ignored, it is not possible for someone else to build the project without creating the class them selves. Even if they were to do that, how are they suppose to know what that class should look like?.

Since you all have access to cloud shell, what about having this class load the JSON file instead? That way the class is always there and it can be documented saying that the client secrete just needs to be downloaded and placed in the resources directory.

Reply:
Thank you for providing this feedback! I was initially using the credentials.json file but since we needed the refresh token, I used the Gmail API tutorial link that I believe Megan had showed you (https://github.com/afsarali273/GmailAPI/blob/master/Gmail/src/main/java/com/gmail/api/GmailAPI.java). I wasn't able to find a way to store the refresh token other than in a hidden file, so I was thinking each of us would store the client id, client secret, and refresh token in a Secrets.java file. 
Is there another way that we can store / access the refresh token?

Reply:
For data like that you would put it in a resource file. For example, having a json file that contained all that information and is loaded at start-up would achieve the same goal you have, but would still allow anyone to compile the project. To make it easier to work with, you could wrap that code in a class, which would help document what should be in the json file.

There is a bigger risk with what you do here. That information is going to be compiled into the binary. That means as long as I get my hands on the binary, I can extract that data from it. Once I have that data, I can make all the calls I want using your API key.

Having that data compiled into the binary also means that in order to update it, you would need to re-build the application compared to swapping out a resource file. While this is less of an issue on a project of this size, imagine if you needed to have multiple instances of this running, all with different credentials.

Reply:
Ah I see, this makes a lot more sense. So I removed the Secrets.java file and instead have a credentials.json file that stores the necessary secret keys (client id, client secret, and refresh token). Then, I use that credentials.json file to load the keys each time a new Gmail API service is created. Thank you so much for your help!